### PR TITLE
[MNG-7818] Removed exclusion of hamcrest from junit 4

### PR DESCRIPTION
**Pull request by lprimak**
Fixes https://issues.apache.org/jira/browse/MNG-7818

Removed exclusion of hamcrest from JUnit 4

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`,
       where you replace `MNG-XXX` and `SUMMARY` with the appropriate JIRA issue.
 - [x] Also format the first line of the commit message like `[MNG-XXX] SUMMARY`.
       Best practice is to use the JIRA issue title in both the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by cstamas** [comment]
Nope, this result in "mixed" dependencies: excerpt from maven-core dep tree:
```
...
[INFO] +- org.hamcrest:hamcrest:jar:2.2:test
[INFO] \- junit:junit:jar:4.13.2:test
[INFO]    \- org.hamcrest:hamcrest-core:jar:1.3:test
```

**Review by cstamas — APPROVED** [review]
This results in "mixed" deps, so is a no go.

**Review by cstamas — CHANGES_REQUESTED** [review]
Sorry, i meant "no go" :smile:

**Comment by lprimak** [comment]
Oh I can fix that :)

**Comment by lprimak** [comment]
Test failures are the "locking" issues, not sure how to clean maven repo in GH actions :(

File: pom.xml [hunk 1]
```diff
@@ -525,20 +526,20 @@ under the License.
...
+        <artifactId>hamcrest-core</artifactId>
+        <version>${hamcrestVersion}</version>
         <scope>test</scope>
       </dependency>
       <dependency>
         <groupId>junit</groupId>
         <artifactId>junit</artifactId>
         <version>${junitVersion}</version>
         <scope>test</scope>
-        <exclusions>
```
Comment (xstefank) [inline pom.xml:536]: Shouldn't you actually keep the exclusion to not pull older version of hamcrest that conflicts with the one you defined on the top level?

**Comment by cstamas** [comment]
> Test failures are the "locking" issues, not sure how to clean maven repo in GH actions :(

You mean the IT MNG-7819 ? As that will cleared up once Resolver 1.9.13 released and updated in Maven.

File: pom.xml (hunk 1 above)
Comment (cstamas) [reply pom.xml:536]: I recommend to look at https://hamcrest.org/JavaHamcrest/distributables#maven-upgrade-example

**Comment by cstamas** [comment]
From test-arq-suite branch JUnit4 ()locally build maven-3.9.x this PR + master shrinkwap-resolver + test-arg-suite:
```
[DEBUG]    junit:junit:jar:4.13.2:test
[DEBUG]       org.hamcrest:hamcrest-core:jar:2.2:test (scope managed from compile) (version managed from 1.3)
[DEBUG]          org.hamcrest:hamcrest:jar:2.2:test (scope managed from compile) (version managed from 2.2)
```

So, what happens, is that by placing hamcrest BEFORE junit in depMgt, the hamcrest will "overrule" the 1.3 from junit, and 2.2 (empty hjar) depends on hamcrest:jar 2.2, so all good.

**Review by cstamas — APPROVED** [review]
Good to go IMHO, we may or we may not wait for resolver 1.9.13 (to get rid of one failing IT)

**Comment by elharo** [comment]
Not sure if this is related but I see some warnigns in this area:

```
[INFO] <<< maven-dependency-plugin:3.6.0:analyze (default-cli) < test-compile @ maven-api-meta <<<
[INFO] 
[INFO] 
[INFO] --- maven-dependency-plugin:3.6.0:analyze (default-cli) @ maven-api-meta ---
[WARNING] Unused declared dependencies found:
[WARNING]    org.junit.jupiter:junit-jupiter-engine:jar:5.9.1:test
[WARNING]    org.hamcrest:hamcrest-core:jar:2.2:test
[INFO] 
```

**Comment by jira-importer** [comment]
Resolve #9102

**Pull request merged**